src/Uno.UI/Helpers/WeakEvents/WeakEventManager.cs

```diff
@@ -3,133 +3,131 @@
+		if (handler is null)
 		{
-			if (!_eventHandlers.TryGetValue(eventName, out List<Subscription>? targets))
-			{
-				targets = new List<Subscription>();
-				_eventHandlers.Add(eventName, targets);
-			}
-			else
-			{
-				targets.RemoveAll(subscription => subscription.Subscriber is { IsAlive: false });
-			}
+			// Allow handler to be nullable to not having to deal with nullability checks in callsites, but this indicates a bug.
+			Debug.Fail($"Got a null event handler for event {eventName}.");
+			return;
+		}
 
-			if (handlerTarget == null)
+		var target = handler.Target;
+		if (target is null)
+		{
+			if (_staticHandlers is not null)
 			{
-				// This event handler is a static method
-				targets.Add(new Subscription(null, methodInfo));
-				return;
+				RemoveFrom(_staticHandlers, eventName, handler);
 			}
-
-			targets.Add(new Subscription(new WeakReference(handlerTarget), methodInfo));
 		}
+		else if (_handlersPerTarget.TryGetValue(target, out var handlersPerEvent))
+		{
+			RemoveFrom(handlersPerEvent, eventName, handler);
+		}
+	}
 
-		private void RemoveEventHandler(string eventName, object? handlerTarget, MethodInfo methodInfo)
+	public void HandleEvent(string eventName)
+	{
+		if (_enumeratingHandlersOfEvent is not null)
 		{
-			if (!_eventHandlers.TryGetValue(eventName, out List<Subscription>? subscriptions))
-				return;
+			_enumeratingHandlersOfEventStack ??= new();
+			_enumeratingHandlersOfEventStack.Push(_enumeratingHandlersOfEvent);
+		}
+		_enumeratingHandlersOfEvent = eventName;
 
-			for (int n = subscriptions.Count - 1; n >= 0; n--)
+		try
+		{
+			foreach (var kvp in _handlersPerTarget)
 			{
-				Subscription current = subscriptions[n];
-
-				if (current.Subscriber != null && !current.Subscriber.IsAlive)
+				if (!kvp.Value.TryGetValue(eventName, out var handlers))
 				{
-					// If not alive, remove and continue
-					subscriptions.RemoveAt(n);
+					// Event handlers not found for this target
 					continue;
 				}
 
-				if (current.Subscriber?.Target == handlerTarget && current.Handler == methodInfo)
+				var count = handlers.Count;
+				for (var i = 0; i < count; i++)
 				{
-					// Found the match, we can break
-					subscriptions.RemoveAt(n);
-					break;
+					handlers[i]();
 				}
 			}
 		}
+		finally
+		{
+			_enumeratingHandlersOfEvent = (_enumeratingHandlersOfEventStack?.TryPop(out var previous) ?? false) ? previous : null;
+		}
+	}
 
-		private readonly struct Subscription : IEquatable<Subscription>
+	private void AddTo(Dictionary<string, List<Action>> handlersPerEvent, string eventName, Action handler)
+	{
+		if (handlersPerEvent.TryGetValue(eventName, out var handlers))
 		{
-			public Subscription(WeakReference? subscriber, MethodInfo handler)
+			if (_enumeratingHandlersOfEvent == eventName)
 			{
-				Subscriber = subscriber;
-				Handler = handler ?? throw new ArgumentNullException(nameof(handler));
+				handlers = [..handlers];
+				handlersPerEvent[eventName] = handlers;
```

Comment:
We could use : https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.collectionsmarshal.getvaluereforadddefault?view=net-8.0#system-runtime-interopservices-collectionsmarshal-getvaluereforadddefault-2(system-collections-generic-dictionary((-0-1))-0-system-boolean@)